pkg/hive/cell/health.go

```diff
@@ -0,0 +1,249 @@
+	if s.LastUpdated.IsZero() {
+		sinceLast = "never"
+	} else {
+		sinceLast = time.Since(s.LastUpdated).String() + " ago"
+	}
+	return fmt.Sprintf("Status{ModuleID: %s, Level: %s, Since: %s, Message: %s, Err: %v}",
+		s.ModuleID, s.Level, sinceLast, s.Message, s.Err)
+}
+
+// NewHealthProvider starts and returns a health status which collects emitted
+// health status declarations and updates module health state.
+func NewHealthProvider() *HealthProvider {
+	p := &HealthProvider{
+		updates:        make(chan Update, updatesBufferSize),
+		moduleStatuses: make(map[string]Status),
+	}
+	p.running.Store(true)
+
+	p.obs, p.emit, p.complete = stream.Multicast[Update]()
+
+	// Listen for updates, use buffered channel to avoid blocking.
+	go func() {
+		for u := range p.updates {
+			prev := func() Status {
+				p.mu.Lock()
+				defer p.mu.Unlock()
+				t := time.Now()
+				prev := p.moduleStatuses[u.ModuleID]
+				ns := Status{
+					Update:      u,
+					LastUpdated: t,
+				}
+				switch u.Level {
+				case StatusOK:
+					ns.LastOK = t
+				case StatusStopped:
+					// If Stopped, set that module was stopped and preserve last known status.
+					ns = prev
+					ns.Stopped = true
+				}
+				p.moduleStatuses[u.ModuleID] = ns
+				log.WithField("status", ns.String()).Debug("Processed new health status")
+				return prev
+			}()
+			p.processed.Add(1)
+			if prev.Stopped {
+				log.Warnf("module %q reported health status after being Stopped", u.ModuleID)
+			}
+		}
+	}()
+
+	return p
+}
+
+const providerDrainPollingInterval = 100 * time.Millisecond
+
+// Finish stops the status provider, and waits for all updates to be processed or
+// returns an error if the context is cancelled first.
+func (p *HealthProvider) Stop(ctx context.Context) error {
+	defer close(p.updates)
+	p.running.Store(false) // following this, no new reporters will send.
+	// Wait for any remaining reporters to finish.
+	p.wg.Wait()
+	// Wait for all updates to be processed, ensuring a consistent state.
+done:
+	for {
+		select {
+		case <-inctimer.After(providerDrainPollingInterval):
+			processed := p.processed.Load()
+			queued := p.queued.Load()
+			// Wait for existing reporters to finish.
+			if processed >= queued {
+				break done
+			}
+		case <-ctx.Done():
+			return fmt.Errorf("failed to drain health status provider: %w", ctx.Err())
+		}
+	}
+	return nil
+}
+
+// forModule returns a module scoped status reporter handle for emitting status updates.
+// This is used to automatically provide declared modules with a status reported.
+func (p *HealthProvider) forModule(moduleID string) StatusReporter {
+	p.mu.Lock()
+	p.moduleStatuses[moduleID] = Status{Update: Update{
+		ModuleID: moduleID,
+		Level:    StatusUnknown,
+		Message:  "No status reported yet"},
+	}
+	p.mu.Unlock()
+
+	return &reporter{
+		moduleID: moduleID,
+		emit: func(u Update) {
+			if !p.running.Load() {
+				return
+			}
+			p.wg.Add(1)
+			defer p.wg.Done()
+			p.queued.Add(1)
+			p.updates <- u
```

Comment:
I don't see the benefit over using a buffered channel and separate worker over just taking a write lock here and updating the module's status directly.

Reply:
I think you're right, a bit of premature optimization here.